workflow-ui/internal-testing-android/src/test/java/com/squareup/workflow1/ui/internal/test/IdlingDIspatcherTest.kt

```diff
@@ -0,0 +1,82 @@
+package com.squareup.workflow1.ui.internal.test
+
+import com.google.common.truth.Truth.assertThat
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.withContext
+import kotlinx.coroutines.yield
+import org.junit.Assert.fail
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+import kotlin.coroutines.ContinuationInterceptor
+import kotlin.coroutines.CoroutineContext
+
+@RunWith(RobolectricTestRunner::class)
+class IdlingDIspatcherTest {
+
+  @Test fun `should not go to idle when dispatching to another dispatcher`() = runBlocking {
+
+    val a = NamedDispatcher("a")
+    val b = NamedDispatcher("b")
+
+    val idler = IdlingDispatcher(a)
+
+    assertThat(idler.isIdle()).isTrue()
+
+    withContext(idler) {
+      assertThat(idler.isIdle()).isFalse()
+      assertThat(coroutineContext[ContinuationInterceptor]).isEqualTo(idler)
+
+      withContext(b) {
+        assertThat(idler.isIdle()).isFalse()
+        assertThat(coroutineContext[ContinuationInterceptor]).isEqualTo(b)
+      }
+
+      assertThat(idler.isIdle()).isFalse()
+      assertThat(coroutineContext[ContinuationInterceptor]).isEqualTo(idler)
+    }
+
+    assertThat(idler.isIdle()).isTrue()
+  }
+
+  @Test fun `should return to idle if a coroutine is cancelled before completion`() = runBlocking {
+
+    val a = NamedDispatcher("a")
+    val b = NamedDispatcher("b")
+
+    val idler = IdlingDispatcher(a)
+
+    // This will never be completed
+    val lock = CompletableDeferred<Unit>()
+
+    assertThat(idler.isIdle()).isTrue()
+
+    val job = launch(idler) {
+      withContext(b) {
+        lock.await()
+        fail("unreachable")
+      }
+    }
+
```

Comment:
Would `isIdle()` be `false` before the `yield()`? Should we add such an assertion?

Reply:
`isIdle()` is `true` before the yield.  The assertion is made before `launch { ... }`.

Unless we launch with `start = UNDISPATCHED` or use `Dispatchers.Unconfined`, the runnable is queued for dispatch.  And since everything is executing on a single thread, it won't actually be dispatched until we call `yield()` or `job.join()`.

Reply:
I updated the test a bit to add clarification.  The short answer to your question, based upon the previous state, is that it wouldn't be false, and then it would be again. :)